document/cast.go

```diff
@@ -0,0 +1,220 @@
+package document
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strconv"
+	"time"
+)
+
+// CastAs casts v as the selected type when possible.
+func (v Value) CastAs(t ValueType) (Value, error) {
+	if v.Type == t {
+		return v, nil
+	}
+
+	// Null values always remain null.
+	if v.Type == NullValue {
+		return v, nil
+	}
+
+	switch t {
+	case BoolValue:
+		return v.CastAsBool()
+	case IntegerValue:
+		return v.CastAsInteger()
+	case DoubleValue:
+		return v.CastAsDouble()
+	case DurationValue:
+		return v.CastAsDuration()
+	case BlobValue:
+		return v.CastAsBlob()
+	case TextValue:
+		return v.CastAsText()
+	case ArrayValue:
+		return v.CastAsArray()
+	case DocumentValue:
+		return v.CastAsDocument()
+	}
+
+	return Value{}, fmt.Errorf("cannot cast %q as %q", v.Type, t)
+}
+
+// CastAsBool casts according to the following rules:
+// Integer: true if truthy, otherwise false.
+// Text: uses strconv.Parsebool to determine the boolean value,
+// it fails if the text doesn't contain a valid boolean.
+// Any other type is considered an invalid cast.
+func (v Value) CastAsBool() (Value, error) {
+	switch v.Type {
+	case BoolValue:
+		return v, nil
+	case IntegerValue:
+		return NewBoolValue(v.V.(int64) != 0), nil
+	case TextValue:
+		b, err := strconv.ParseBool(string(v.V.([]byte)))
+		if err != nil {
+			return Value{}, err
+		}
+		return NewBoolValue(b), nil
+	}
+
+	return Value{}, fmt.Errorf("cannot cast %q as bool", v.Type)
+}
+
+// CastAsInteger casts according to the following rules:
+// Bool: returns 1 if true, 0 if false.
+// Double: cuts off the decimal and remaining numbers.
+// Text: uses strconv.ParseFloat to determine the double value,
+// then casts it to an integer. it fails if the text doesn't
+// contain a valid float value.
+// Any other type is considered an invalid cast.
+func (v Value) CastAsInteger() (Value, error) {
+	switch v.Type {
+	case IntegerValue:
+		return v, nil
+	case BoolValue:
+		if v.V.(bool) {
+			return NewIntegerValue(1), nil
+		}
+		return NewIntegerValue(0), nil
+	case DoubleValue:
+		return NewIntegerValue(int64(v.V.(float64))), nil
+	case TextValue:
+		f, err := strconv.ParseFloat(string(v.V.([]byte)), 64)
```

Comment:
It's because we want to allow this:

```
genji> SELECT CAST('10.5' AS integer);
{
  "CAST('10.5' AS integer)": 10
}
```

Parsing as int would make this case fail